flox/xarray.py

```diff
@@ -296,10 +303,12 @@ def xarray_reduce(
                     f"Please provided bin edges for group variable {idx} "
                     f"named {group_names[idx]} in expected_groups."
                 )
-            expected_groups[idx] = _get_expected_groups(b.data, sort=sort, raise_if_dask=True)
+            expected_groups[idx] = _get_expected_groups(b_.data, sort=sort, raise_if_dask=True)
 
-    expected_groups = _convert_expected_groups_to_index(expected_groups, isbin, sort=sort)
-    group_shape = tuple(len(e) for e in expected_groups)
+    expected_groups = _convert_expected_groups_to_index(expected_groups, isbins, sort=sort)
+    # TODO: _convert_expected_groups_to_index can return None which is not good
```

Comment:
`expected_groups` cannot have a `None` element at this stage see:
```
expected_groups[idx] = _get_expected_groups(b_.data, sort=sort, raise_if_dask=True)
```

This may be complicated from a typing perspective.

Reply:
`_get_expected_groups` can also return `None` so it's not so easy to untangle this.
And even if `expected_groups` was narrowed properly it doesn't matter because `_convert_expected_groups_to_index` still has the None in it's return type.

```python
def test2(a: tuple[str | int, ...]) -> tuple[str | int, ...]:
    return a


b: tuple[int, ...] = (1, 2)
reveal_type(test2(a=b))  # note: Revealed type is "builtins.tuple[Union[builtins.str, builtins.int], ...]"
```

There may not be logic bug here but this part of the code is really hard to understand and could do with a little simplification.

Reply:
:+1: Agreed. One simplification would be to remove the `raise_if_dask` kwarg. It's only set to False in one place, we can explicitly skip it there.

Typing `_convert_expected_groups_to_index` is hard because it handles some very flexible user input but happy to hear suggestions.

Reply:
I moved similar stuff inside the loop which simplified it a little [e73f6e8](https://github.com/xarray-contrib/flox/pull/144/commits/e73f6e83df5e38f611a6fd6099a63f1b8f2c1b81)
`group_names` can probably be replaced by `group_sizes` as well.